service/authorization/authorization.go

```diff
@@ -186,183 +185,215 @@ func (as *AuthorizationService) GetDecisions(ctx context.Context, req *connect.R
-				//
-				// TODO: we should already have the subject mappings here and be able to just use OPA to trim down the known data attr values to the ones matched up with the entities
-				//
-				entities := ec.GetEntities()
-				req := connect.Request[authorization.GetEntitlementsRequest]{
-					Msg: &authorization.GetEntitlementsRequest{
-						Entities: entities,
-						Scope:    &allPertinentFqnsRA,
-					},
-				}
+		attrDefs, err = populateAttrDefValueFqns(attrDefs)
+		if err != nil {
+			return nil, connect.NewError(connect.CodeInternal, err)
+		}
+		allPertinentFQNS.AttributeValueFqns = append(allPertinentFQNS.GetAttributeValueFqns(), ra.GetAttributeValueFqns()...)
 
-				auditECEntitlements := make([]audit.EntityChainEntitlement, 0)
-				auditEntityDecisions := make([]audit.EntityDecision, 0)
+		// get the relevant resource attribute fqns
+		for _, attrDef := range attrDefs {
+			if attrDef.GetRule() == policy.AttributeRuleTypeEnum_ATTRIBUTE_RULE_TYPE_ENUM_HIERARCHY {
+				for _, value := range attrDef.GetValues() {
+					allPertinentFQNS.AttributeValueFqns = append(allPertinentFQNS.AttributeValueFqns, value.GetFqn())
+				}
+			}
+		}
+		attrDefsReqs = append(attrDefsReqs, attrDefs)
+		attrValsReqs = append(attrValsReqs, attrVals)
+		fqnsReqs = append(fqnsReqs, fqns)
+	}
 
-				// Entitlements for environment entites in chain
-				envEntityAttrValues := make(map[string][]string)
-				// Entitlements for sbuject entities in chain
-				subjectEntityAttrValues := make(map[string][]string)
+	var ecChainEntitlementsResponse []*connect.Response[authorization.GetEntitlementsResponse]
+	for _, ec := range dr.GetEntityChains() {
+		entities := ec.GetEntities()
+		if len(entities) == 0 {
+			ecChainEntitlementsResponse = append(ecChainEntitlementsResponse, nil)
+			continue
+		}
+		req := connect.Request[authorization.GetEntitlementsRequest]{
+			Msg: &authorization.GetEntitlementsRequest{
+				Entities: entities,
+				Scope:    allPertinentFQNS,
+			},
+		}
+		ecEntitlements, err := as.GetEntitlements(ctx, &req)
+		if err != nil {
+			// TODO: should all decisions in a request fail if one entity entitlement lookup fails?
+			return nil, db.StatusifyError(err, db.ErrTextGetRetrievalFailed, slog.String("extra", "getEntitlements request failed"))
+		}
+		ecChainEntitlementsResponse = append(ecChainEntitlementsResponse, ecEntitlements)
+	}
 
-				//nolint:nestif // handle empty entity / attr list
-				if len(entities) == 0 || len(allPertinentFqnsRA.GetAttributeValueFqns()) == 0 {
-					as.logger.WarnContext(ctx, "empty entity list and/or entity data attribute list")
-				} else {
-					ecEntitlements, err := as.GetEntitlements(ctx, &req)
-					if err != nil {
-						// TODO: should all decisions in a request fail if one entity entitlement lookup fails?
-						return nil, db.StatusifyError(err, db.ErrTextGetRetrievalFailed, slog.String("extra", "getEntitlements request failed"))
+	for reqIdx, ra := range dr.GetResourceAttributes() {
```

Comment:
nit: similar here, maybe `raIdx` so it doesnt get confused with iterating through multiple decision requests